### PR TITLE
o1vm/pickles: introduce the computation of the quotient polynomial

### PR DESCRIPTION
**Pull request by dannywillems**
TODO:
- absorb the commitments
- add in the opening proof

Note that this will break the e2e tests because selectors polynomials are not defined correctly at the moment.

### CODE REVIEW

File: o1vm/src/pickles/prover.rs [hunk 1]
```diff
@@ -24,6 +24,20 @@ use super::{
...
 use crate::E;
+use thiserror::Error;
+
+/// Errors that can arise when creating a proof
+#[derive(Error, Debug, Clone)]
+pub enum ProverError {
+    #[error("the proof could not be constructed: {0}")]
+    Generic(&'static str),
+
+    #[error("the provided (witness) constraints was not satisfied: {0}")]
```
Comment (marcbeunardeau88) [inline o1vm/src/pickles/prover.rs:35]: either the constraints were not satisfied, or the witness satisfying

File: o1vm/src/pickles/prover.rs [hunk 2]
```diff
@@ -24,6 +24,20 @@ use super::{
...
+    ConstraintNotSatisfied(String),
+
+    #[error("the provided (witness) constraint has degree {0} > allowed {1}; expr: {2}")]
```
Comment (marcbeunardeau88) [inline o1vm/src/pickles/prover.rs:38]:
```suggestion
    #[error("the provided constraint has degree {0} > allowed {1}; expr: {2}")]
```

File: o1vm/src/pickles/prover.rs (hunk 1 above)
Comment (marcbeunardeau88) [inline o1vm/src/pickles/prover.rs:35]:
```suggestion
    #[error("the provided constraint was not satisfied: {0}")]
```

File: o1vm/src/pickles/prover.rs [hunk 3]
```diff
@@ -176,7 +190,72 @@ where
...
+        let mut last_constraint_failed = None;
+        // Only for debugging purposes
+        for expr in constraints.iter() {
+            let fail_q_division =
+                ProverError::ConstraintNotSatisfied(format!("Unsatisfied expression: {:}", expr));
+            // Check this expression are witness satisfied
+            let (_, res) = expr
+                .evaluations(&column_env)
+                .interpolate_by_ref()
+                .divide_by_vanishing_poly(domain.d1)
```
Comment (marcbeunardeau88) [inline o1vm/src/pickles/prover.rs:203]: are we paying for a division for each constraints just to check, and re doing that division line 246 ?

File: o1vm/src/pickles/prover.rs [hunk 4]
```diff
@@ -176,7 +190,72 @@ where
...
+        // must be equal to 0 as the constraints polynomial and Z_H(X) are both
+        // equal on H.
+        if !res.is_zero() {
+            fail_final_q_division();
+        }
+
+        quotient
+    };
+
+    let _t_comm = srs.commit_non_hiding(&quotient_poly, 7);
```
Comment (marcbeunardeau88) [inline o1vm/src/pickles/prover.rs:258]: we should absorb the t before squeezing the next FS challenge !

Comment (dannywillems) [reply o1vm/src/pickles/prover.rs:258]: It is in the future (draft) PR, see https://github.com/o1-labs/proof-systems/pull/2655. It is also written in the description of this PR that it is a TODO.

File: o1vm/src/pickles/prover.rs (hunk 3 above)
Comment (dannywillems) [reply o1vm/src/pickles/prover.rs:203]: Yes, I am adding an option soon to activate only this in a "debug" mode. PR coming.

Comment (marcbeunardeau88) [reply o1vm/src/pickles/prover.rs:203]: I'd remove that, doesn't it pollute the code ?
Note that we can check the reminder when dividing by the vanishing poly if we really want a check

Comment (dannywillems) [reply o1vm/src/pickles/prover.rs]: I did remove https://github.com/o1-labs/proof-systems/pull/2652/commits/8e87244a98d55b90d175ad389611a3c98bd16b34 and https://github.com/o1-labs/proof-systems/pull/2652/commits/96d42c127ef025869c91e5fed680e0e383108706 as you seem to have a strong opinion on this. I would probably keep the commits somewhere on my side for debugging purposes when finishing the prover.

File: o1vm/src/pickles/prover.rs [hunk 5]
```diff
@@ -176,7 +184,48 @@ where
         }
     };
 
-    // FIXME: add quotient polynomial
+    let quotient_poly: DensePolynomial<G::ScalarField> = {
+        // Compute ∑ α^i constraint_i as an expression
+        let combined_expr =
+            E::combine_constraints(0..(constraints.len() as u32), (constraints).to_vec());
```
Comment (marcbeunardeau88) [inline o1vm/src/pickles/prover.rs:190]: not a big deal, but I'd rather have this computed elsewhere
The expressions should not be computed at proving time, for separation of concept

File: o1vm/src/pickles/prover.rs [hunk 6]
```diff
@@ -176,7 +184,48 @@ where
...
+
+        // And we interpolate using the evaluations
+        let expr_evaluation_interpolated = expr_evaluation.interpolate();
+
+        let fail_final_q_division = || {
+            panic!("Division by vanishing poly must not fail at this point, we checked it before")
+        };
+        // We compute the polynomial t(X) by dividing the constraints polynomial
+        // by the vanishing polynomial, i.e. Z_H(X).
+        let (quotient, res) = expr_evaluation_interpolated
```
Comment (marcbeunardeau88) [inline o1vm/src/pickles/prover.rs:215]: shouldn't res be called rem for remainder ?

File: o1vm/src/pickles/prover.rs [hunk 7]
```diff
@@ -176,7 +184,48 @@ where
...
+            .divide_by_vanishing_poly(domain.d1)
+            .unwrap_or_else(fail_final_q_division);
```
Comment (marcbeunardeau88) [inline o1vm/src/pickles/prover.rs:217]: Why do we have twice the same error for two different fails ?
Also (not a  part of this PR) I don't see the fail happening in divide by vanishing poly, mb it should not returns an option

**Review by marcbeunardeau88 — APPROVED** [review]

**Pull request merged**